src/Views/LocationPanel.vala

```diff
@@ -118,45 +114,41 @@ public class SecurityPrivacy.LocationPanel : Granite.SimpleSettingsPage {
         }
     }
 
-    private void populate_app_listbox () {
-        load_remembered_apps ();
+    private Gtk.Widget create_widget_func (Object object) {
+        var app_permission = (AppPermission) object;
 
-        foreach (var row in listbox.get_children ()) {
-            listbox.remove (row);
-        }
+        var app_row = new LocationRow (app_permission);
 
-        foreach (var app in remembered_apps) {
-            string app_id = app.get_child_value (0).get_string ();
-            bool authed = app.get_child_value (1).get_variant ().get_child_value (0).get_boolean ();
-            var app_info = new DesktopAppInfo (app_id + ".desktop");
+        app_row.notify["authed"].connect (() => {
+            var tuple_vals = new Variant[2];
+            tuple_vals[0] = new Variant.boolean (app_row.authed);
+            tuple_vals[1] = new Variant.uint32 (app_permission.level);
 
-            var app_row = new LocationRow (app_info, authed);
+            var remembered_apps_dict = new VariantDict (location_settings.get_value ("remembered-apps"));
+            remembered_apps_dict.insert_value (app_permission.id, new Variant.tuple (tuple_vals));
 
-            app_row.active_changed.connect ((active) => {
-                uint32 level = get_app_level (app_id);
-                save_app_settings (app_id, active, level);
-            });
+            location_settings.set_value ("remembered-apps", remembered_apps_dict.end ());
+        });
 
-            listbox.add (app_row);
-        }
+        return app_row;
     }
 
     private void load_remembered_apps () {
-        remembered_apps = location_settings.get_value ("remembered-apps");
-        remembered_apps_dict = new VariantDict (location_settings.get_value ("remembered-apps"));
-    }
+        liststore.remove_all ();
 
-    private void save_app_settings (string desktop_id, bool authorized, uint32 accuracy_level) {
-        Variant[] tuple_vals = new Variant[2];
-        tuple_vals[0] = new Variant.boolean (authorized);
-        tuple_vals[1] = new Variant.uint32 (accuracy_level);
-        remembered_apps_dict.insert_value (desktop_id, new Variant.tuple (tuple_vals));
-        location_settings.set_value ("remembered-apps", remembered_apps_dict.end ());
-        load_remembered_apps ();
-    }
+        foreach (var app in location_settings.get_value ("remembered-apps")) {
+            var app_permission = new AppPermission (
+                app.get_child_value (0).get_string (),
+                app.get_child_value (1).get_variant ().get_child_value (0).get_boolean (),
+                app.get_child_value (1).get_variant ().get_child_value (1).get_uint32 ()
+            );
```

Comment:
Same here, it would be better to have

```suggestion
            string desktop_id;
            bool authed;
            uint32 accuracy_level;
            app.get ("s(bu)", out desktop_id, out authed, out accuracy_level)
            var app_permission = new AppPermission (desktop_id, authed, accuracy_level);
```

Reply:
Hm I get `'s(bu)' is not a valid GVariant format string`